### PR TITLE
Pseudo PR - community review for initial release

### PR DESCRIPTION
**Pull request by ewels**
> _Do not merge - pseudo PR to give review interface for entire pipeline_

Community review for initial release. Aim is to get at least two PR approvals from people in the nf-core community who are not the main developers of this pipeline ✅ 

https://nf-co.re/developers/adding_pipelines#making-the-first-release

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,118 @@
...
+3. [Running the pipeline](docs/usage.md)
+4. [Output and how to interpret the results](docs/output.md)
+5. [Troubleshooting](https://nf-co.re/usage/troubleshooting)
+
+## Usage
+
+### With a samples.csv file
+
+```bash
+nextflow run nf-core/kmermaid --outdir s3://olgabot-maca/nf-kmer-similarity/ --samples samples.csv
```
Comment (maxulysse) [inline README.md]:
```suggestion
nextflow run nf-core/kmermaid --samples samples.csv
```

The --outdir seems to be very AWS specific to me

Comment (maxulysse) [reply README.md]: same with all s3 path in the README

File: README.md [hunk 2]
```diff
@@ -1 +1,118 @@
...
+
+```bash
+nextflow run nf-core/kmermaid --outdir s3://olgabot-maca/nf-kmer-similarity/ \
+  --fastas '*.fasta'
+```
+
+### With bam file
+
+```bash
+nextflow run czbiohub/nf-kmer-similarity --outdir s3://olgabot-maca/nf-kmer-similarity/ \
```
Comment (maxulysse) [inline README.md]: pipeline name should be nf-core/kmermaid

File: README.md [hunk 3]
```diff
@@ -1 +1,118 @@
...
+  --bam 'possorted_genome_bam.bam'
+```
+
+### With split kmer
+
+```bash
+nextflow run czbiohub/nf-kmer-similarity --outdir s3://olgabot-maca/nf-kmer-similarity/ --samples samples.csv --split_kmer --subsample 1000
```
Comment (maxulysse) [inline README.md]: pipeline name should be nf-core/kmermaid

File: bin/markdown_to_html.r [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env Rscript
```
Comment (maxulysse) [inline bin/markdown_to_html.r]: I think this file can be remove as the `.py` file is used now

File: conf/awsbatch.config [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (maxulysse) [inline conf/awsbatch.config]: This file can be removed too

File: docs/configuration/adding_your_own.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
+# nf-core/nf-kmer-similarity: Configuration for other clusters
```
Comment (maxulysse) [inline docs/configuration/adding_your_own.md]: This file can be removed too

File: docs/installation.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
+# nf-core/nf-kmer-similarity: Installation
```
Comment (maxulysse) [inline docs/installation.md]: This file can be removed as well

**Comment by ewels** [comment]
Before a more in-depth look at the code, here are a few things that need resolving from a first glance:

### Template update

You’re a bit behind the main nf-core template. You created the `TEMPLATE` branch with version `1.9` of nf-core/tools and we’re on version `1.10.2` (about to release `1.10.3`). The new minor patch release might come very soon so hopefully you’ll get an automated sync PR if that happens.

Either way, you need to update the `TEMPLATE` branch with the latest version and then bring those updates across to the main pipeline code. The update should hopefully not be too tough, mostly additions. For example, you're missing a bunch of GitHub actions files ([kmermaid](https://github.com/nf-core/kmermaid/tree/dev/.github/workflows) vs [template](https://github.com/nf-core/tools/tree/master/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/.github/workflows)) which will be added.

### Test warnings

Generally the tests are looking great - no failures ✅ , however there are a few warnings. These are mostly about updating a bunch of conda packages (no need to do all of these, but some look like they should be easy). The others should be fixed by the template update.

**`nf-core lint` overall result: Passed with warnings ❗**

Posted for pipeline commit e1f736c

```diff
+| ✅ 109 tests passed       |+
!| ❗  21 tests had warnings |!
-| ❌   0 tests failed       |-
```

<details>

<summary>All lint test results</summary>

### :heavy_exclamation_mark: Test warnings:

* [Test #1](https://nf-co.re/errors#1) - File not found: `.github/workflows/awstest.yml`
* [Test #1](https://nf-co.re/errors#1) - File not found: `.github/workflows/awsfulltest.yml`
* [Test #6](https://nf-co.re/errors#6) - Found a bioconda environment.yml file but no badge in the README
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::python=3.7.3`, `3.9.0` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::markdown=3.1.1`, `3.3.2` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::pymdown-extensions=6.0`, `8.0.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::pygments=2.5.2`, `2.7.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::tqdm=4.43.0`, `4.50.2` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::gxx_linux-64=7.3.0`, `9.3.0` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `conda-forge::s3fs=0.4.2`, `0.5.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `bioconda::samtools=1.10`, `1.11` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `bioconda::pysam=0.16.0`, `0.16.0.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `fastp=0.20.0`, `0.20.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `matplotlib=3.1.1`, `3.3.2` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `multiqc=1.8`, `1.9` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `numpy=1.17.5`, `1.19.2` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `pathos=0.2.5`, `0.2.6` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `pip=20.0.2`, `20.2.4` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `pytest=5.3.4`, `6.1.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `sphinx=2.3.1`, `3.2.1` available
* [Test #8](https://nf-co.re/errors#8) - Conda package is not latest available: `sortmerna=2.1b`, `4.2.0` available

### :white_check_mark: Tests passed:

* [Test #1](https://nf-co.re/errors#1) - File found: `nextflow.config`
* [Test #1](https://nf-co.re/errors#1) - File found: `nextflow_schema.json`
* [Test #1](https://nf-co.re/errors#1) - File found: `Dockerfile`
* [Test #1](https://nf-co.re/errors#1) - File found: `LICENSE` or `LICENSE.md` or `LICENCE` or `LICENCE.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `README.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `CHANGELOG.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `docs/README.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `docs/output.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `docs/usage.md`
* [Test #1](https://nf-co.re/errors#1) - File found: `.github/workflows/branch.yml`
* [Test #1](https://nf-co.re/errors#1) - File found: `.github/workflows/ci.yml`
* [Test #1](https://nf-co.re/errors#1) - File found: `.github/workflows/linting.yml`
* [Test #1](https://nf-co.re/errors#1) - File found: `main.nf`
* [Test #1](https://nf-co.re/errors#1) - File found: `environment.yml`
* [Test #1](https://nf-co.re/errors#1) - File found: `conf/base.config`
* [Test #1](https://nf-co.re/errors#1) - File not found check: `Singularity`
* [Test #1](https://nf-co.re/errors#1) - File not found check: `parameters.settings.json`
* [Test #1](https://nf-co.re/errors#1) - File not found check: `.travis.yml`
* [Test #3](https://nf-co.re/errors#3) - Licence check passed
* [Test #2](https://nf-co.re/errors#2) - Dockerfile check passed
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.name`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.nextflowVersion`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.description`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.version`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.homePage`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `timeline.enabled`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `trace.enabled`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `report.enabled`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `dag.enabled`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `process.cpus`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `process.memory`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `process.time`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `params.outdir`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `params.input`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `manifest.mainScript`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `timeline.file`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `trace.file`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `report.file`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `dag.file`
* [Test #4](https://nf-co.re/errors#4) - Config variable found: `process.container`
* [Test #4](https://nf-co.re/errors#4) - Config variable (correctly) not found: `params.version`
* [Test #4](https://nf-co.re/errors#4) - Config variable (correctly) not found: `params.nf_required_version`
* [Test #4](https://nf-co.re/errors#4) - Config variable (correctly) not found: `params.container`
* [Test #4](https://nf-co.re/errors#4) - Config variable (correctly) not found: `params.singleEnd`
* [Test #4](https://nf-co.re/errors#4) - Config variable (correctly) not found: `params.igenomesIgnore`
* [Test #4](https://nf-co.re/errors#4) - Config `timeline.enabled` had correct value: `true`
* [Test #4](https://nf-co.re/errors#4) - Config `report.enabled` had correct value: `true`
* [Test #4](https://nf-co.re/errors#4) - Config `trace.enabled` had correct value: `true`
* [Test #4](https://nf-co.re/errors#4) - Config `dag.enabled` had correct value: `true`
* [Test #4](https://nf-co.re/errors#4) - Config `manifest.name` began with `nf-core/`
* [Test #4](https://nf-co.re/errors#4) - Config variable `manifest.homePage` began with https://github.com/nf-core/
* [Test #4](https://nf-co.re/errors#4) - Config `dag.file` ended with `.svg`
* [Test #4](https://nf-co.re/errors#4) - Config variable `manifest.nextflowVersion` started with >= or !>=
* [Test #4](https://nf-co.re/errors#4) - Config `process.container` looks correct: `nfcore/kmermaid:dev`
* [Test #4](https://nf-co.re/errors#4) - Config `manifest.version` ends in `dev`: `'1.0.0dev'`
* [Test #5](https://nf-co.re/errors#5) - GitHub Actions 'branch' workflow is triggered for PRs to master: `./.github/workflows/branch.yml`
* [Test #5](https://nf-co.re/errors#5) - GitHub Actions 'branch' workflow looks good: `./.github/workflows/branch.yml`
* [Test #5](https://nf-co.re/errors#5) - GitHub Actions CI is triggered on expected events: `./.github/workflows/ci.yml`
* [Test #5](https://nf-co.re/errors#5) - CI is building the correct docker image: `docker build --no-cache . -t nfcore/kmermaid:dev`
* [Test #5](https://nf-co.re/errors#5) - CI is pulling the correct docker image: docker pull nfcore/kmermaid:dev
* [Test #5](https://nf-co.re/errors#5) - CI is tagging docker image correctly: docker tag nfcore/kmermaid:dev nfcore/kmermaid:dev
* [Test #5](https://nf-co.re/errors#5) - Continuous integration checks minimum NF version: `./.github/workflows/ci.yml`
* [Test #5](https://nf-co.re/errors#5) - GitHub Actions linting workflow is triggered on PR and push: `./.github/workflows/linting.yml`
* [Test #5](https://nf-co.re/errors#5) - Continuous integration runs Markdown lint Tests: `./.github/workflows/linting.yml`
* [Test #5](https://nf-co.re/errors#5) - Continuous integration runs nf-core lint Tests: `./.github/workflows/linting.yml`
* [Test #6](https://nf-co.re/errors#6) - README Nextflow minimum version badge matched config. Badge: `20.07.1`, Config: `20.07.1`
* [Test #8](https://nf-co.re/errors#8) - Conda environment name was correct (nf-core-kmermaid-1.0.0dev)
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::python=3.7.3`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::markdown=3.1.1`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::pymdown-extensions=6.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::pygments=2.5.2`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::tqdm=4.43.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::gxx_linux-64=7.3.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `conda-forge::s3fs=0.4.2`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `bioconda::sourmash=3.5.0`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `bioconda::sourmash=3.5.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `bioconda::samtools=1.10`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `bioconda::screed=1.0.4`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `bioconda::screed=1.0.4`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `bioconda::khmer=3.0.0a3`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `bioconda::khmer=3.0.0a3`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `bioconda::pysam=0.16.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `anaconda::make=4.2.1`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `anaconda::make=4.2.1`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `alabaster=0.7.12`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `alabaster=0.7.12`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `fastp=0.20.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `fastqc=0.11.9`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `fastqc=0.11.9`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `matplotlib=3.1.1`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `multiqc=1.8`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `numpy=1.17.5`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `pathos=0.2.5`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `pip=20.0.2`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `pytest=5.3.4`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `seqtk=1.3`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `seqtk=1.3`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `ska=1.0`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `ska=1.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `sphinx=2.3.1`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `jupyter=1.0.0`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `jupyter=1.0.0`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `sortmerna=2.1b`
* [Test #8](https://nf-co.re/errors#8) - Conda dependency had pinned version number: `ripgrep=12.1.1`
* [Test #8](https://nf-co.re/errors#8) - Conda package is latest available: `ripgrep=12.1.1`
* [Test #8](https://nf-co.re/errors#8) - Pip dependency had pinned version number: bam2fasta==1.0.8
* [Test #8](https://nf-co.re/errors#8) - PyPi package is latest available: 1.0.8
* [Test #8](https://nf-co.re/errors#8) - Pip dependency had pinned version number: sencha==1.0.3
* [Test #8](https://nf-co.re/errors#8) - PyPi package is latest available: 1.0.3

### Run details:

* nf-core/tools version 1.11.dev0
* Run at `2020-10-22 09:36:48`

</details>

### Files that shouldn't be there

I can see a few files that can be / should be removed I think:

* `Makefile`
    * Seems to be for pushing the docker image to dockerhub. This is entirely automated with GitHub actions now.
* `environment_hulk.yml`
    * Seems to be specific to CZI? eg. has [hardcoded paths](https://github.com/nf-core/kmermaid/blob/e1f736c5c694037d548af53e8d1d1cacf120327f/environment_hulk.yml#L9).
* `testing` directorey
    * Seems to have minimal test datasets. These should be in the `nf-core/test-datasets` repo instead (there's [already a branch for kmermaid](https://github.com/nf-core/test-datasets/tree/kmermaid))
* `bin/markdown_to_html.r` - this was replaced by the template python version and is no longer used

**Review by maxulysse — COMMENTED** [review]
Switching to a single comment instead of commenting into files, not to spam everyone.

- `.gitignore` seems to be a bit overkill to me.
- `Dockerfile` I guess that the last lines are to test that tools are working. Is it really necessary?
- `Makefile` I don't think it's actually useful in the repo
- `README`:
  - There are still some `TODO` comments left.
  - Usage should be made less specific
  - I think you can already set up the Zenodo since you have a tag
- There is two `assets/email_template` file, not sure which one we should keep `html` or `txt`
- `assets/rrna-db-defaults.txt` what is this file used for?
- Remove ` bin/markdown_to_html.r`, ` conf/awsbatch.config`, `docs/configuration/adding_your_own.md`, `docs/installation.md`, `docs/troubleshooting.md`
- `docker/sysctl.conf` what is this file used for?
- params
  - `--csv_pairs` and `--csv_singles` could both be replaced by a single `--input`. since you have a header, you could check the number of column to asses if single or double.
  - `--fastas` and `--protein_fastas` why plural?
  - `--removeRiboRNA`, `--saveNonRiboRNAReads`, `--rNA_database_manifest` are not snake_cased.
  - `--molecule` or `--molecules`, I've seen both in the docs and configs so far.
  - `--save_fastas` should be `--save_fasta_dir`
- Why the `environment_hulk.yml` file?
- `nextflow.config` not up to date with `TEMPLATE`
- there should not be any testing data in the repo, but in the `nf-core/test-datasets` one, on the `kmermaid` branch

Apart from these tiny comments, I am quite happy with the pipeline, I think you're mainly missing updates from the TEMPLATE, and maybe some input params improvement.

**Comment by maxulysse** [comment]
Thanks for your reply and comments.
Just replying to ones that need to, as you were very clear on all the other comments.

> > * `.gitignore` seems to be a bit overkill to me.
> 
> Not sure that this is a major problem? There are Python files in the `bin` and when I edit with PyCharm, then excess files get created. It seems to be fine to specify all the possible files that may want to be ignored, but open to suggestions!

Not a problem at all, I agree, it was more a remark.

> 
> > * `Dockerfile` I guess that the last lines are to test that tools are working. Is it really necessary?
> 
> Yes, it helps to make sure everything got installed properly in the Docker image. I've had issues with the Docker image even working in the first place without having those checks in the past.

OK, then I have no issue keeping that, I'm pretty sure that with DSL2 we're moving out of Dockerfiles anyway

> > * `Makefile` I don't think it's actually useful in the repo
> 
> This speeds up my local workflow to build the docker image locally and not have to look up the `docker build` command every time. I can remove the `docker push` rule if that would help.

You said it yourself: `This speeds up my local workflow`, I don't think it belongs here.

> > * There is two assets/email_template file, not sure which one we should keep html or txt
>
> According to the latest nf-core/tools template, both are needed? https://github.com/nf-core/tools/tree/master/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/assets

Totally right, my bad...

> > * Why the `environment_hulk.yml` file?
> 
> We have a local machine that needs a lot of love and care when installing `kmermaid` so we made it a special environment file.

You said it yourself: `We have a local machine`, I don't think it belongs here.

BTW, we do have a `Hulk` as well at NGI ;-)

> > * `nextflow.config` not up to date with `TEMPLATE`
> 
> These PRs didn't do it?
> 
>     * #93
> 
>     * #110
> 

At least these lines are not in your config file:
https://github.com/nf-core/tools/blob/f14c7a5692c96896c2ed1b730bc355093dca2247/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/nextflow.config#L82-L87

(That was what make me thinking that it might not have been up to date to start with)

I will have a more in depth look to make sure

> > Apart from these tiny comments, I am quite happy with the pipeline, I think you're mainly missing updates from the TEMPLATE, and maybe some input params improvement.
> 
> :tada: Thanks! Glad to hear! :tada:

Looking forward for the big release!!!

**Comment by ewels** [comment]
Regarding the template merge, I guessed you used version 1.9 because of this line: https://github.com/nf-core/kmermaid/blob/7907ca5dc827fd69895a6aeea8d17c39f72184d7/Dockerfile#L1

But if you don't think that you did, then it probably warrants a closer comparison between the template files and the pipeline files...

+1 for removing institute specific stuff. I know it's a hassle, we had to do similar things for nearly all of our pipelines when we ported them to be nf-core instead of SciLifeLab (it wasn't immediate, an `uppmax` config profile shipped with all pipelines for a long time). But this is really the heart of nf-core: that pipelines are generalised as much as is possible so as to be first-class pipelines for anyone anywhere.

**Comment by ewels** [comment]
Stupid _"Close with comment"_ button is too close to the _"Comment"_ button, sorry..

**Review by riturajjain2000 — APPROVED** [review]

**Review by Surajv311 — APPROVED** [review]

**Review by akshitagupta15june — APPROVED** [review]

**Review by abhi18av — APPROVED** [review]

**Review by JayantGoel001 — APPROVED** [review]

**Review by ArnavBalyan — APPROVED** [review]

**Pull request merged**